cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -45,67 +49,109 @@ constexpr static inline const float kAdjustCentersWeight = 7.0f;
  * @param metric
  * @param stream
  * @param mr (optional) memory resource to use for temporary allocations
  */
+template <typename IdxT, typename LabelT>
 inline void predict_float_core(const handle_t& handle,
                                const float* centers,
                                uint32_t n_clusters,
                                uint32_t dim,
                                const float* dataset,
+                               const float* dataset_norm,
                                size_t n_rows,
-                               uint32_t* labels,
+                               LabelT* labels,
                                raft::distance::DistanceType metric,
                                rmm::cuda_stream_view stream,
                                rmm::mr::device_memory_resource* mr)
 {
-  rmm::device_uvector<float> distances(n_rows * n_clusters, stream, mr);
-
-  float alpha;
-  float beta;
   switch (metric) {
-    case raft::distance::DistanceType::InnerProduct: {
-      alpha = -1.0;
-      beta  = 0.0;
-    } break;
     case raft::distance::DistanceType::L2Expanded:
-    case raft::distance::DistanceType::L2Unexpanded: {
-      rmm::device_uvector<float> sqsum_centers(n_clusters, stream, mr);
-      rmm::device_uvector<float> sqsum_data(n_rows, stream, mr);
-      utils::dots_along_rows(n_clusters, dim, centers, sqsum_centers.data(), stream);
-      utils::dots_along_rows(n_rows, dim, dataset, sqsum_data.data(), stream);
-      utils::outer_add(
-        sqsum_data.data(), n_rows, sqsum_centers.data(), n_clusters, distances.data(), stream);
-      alpha = -2.0;
-      beta  = 1.0;
-    } break;
-    // NB: update the description of `knn::ivf_flat::build` when adding here a new metric.
-    default: RAFT_FAIL("The chosen distance metric is not supported (%d)", int(metric));
+    case raft::distance::DistanceType::L2SqrtExpanded: {
+      // todo(lsugy): pass
+      auto workspace = raft::make_device_vector<char, IdxT>(handle, (sizeof(int)) * n_rows);
+
+      raft::distance::MinAndDistanceReduceOp<LabelT, float> redOp;
+      raft::distance::KVPMinReduce<LabelT, float> pairRedOp;
+
+      auto minClusterAndDistance =
+        raft::make_device_vector<cub::KeyValuePair<LabelT, float>, IdxT>(handle, n_rows);
+      cub::KeyValuePair<LabelT, float> initial_value(0, std::numeric_limits<float>::max());
+      thrust::fill(handle.get_thrust_policy(),
+                   minClusterAndDistance.data_handle(),
+                   minClusterAndDistance.data_handle() + minClusterAndDistance.size(),
+                   initial_value);
+
+      auto centroidsNorm = raft::make_device_vector<float, IdxT>(handle, n_clusters);
+      raft::linalg::rowNorm<float, IdxT>(
+        centroidsNorm.data_handle(), centers, dim, n_clusters, raft::linalg::L2Norm, true, stream);
+
+      // todo(lsugy): should fusedL2NN take IdxT!=LabelT?
+      raft::distance::fusedL2NN<float, cub::KeyValuePair<LabelT, float>, LabelT>(
```

Comment:
Probably a comment for #795: shouldn't we have a specialization for fusedLNN that selects only the label? That would simplify things here

Reply:
I'm not sure if it's that easy to implement, since you can't really define a reduction operation between two labels without a distance.